src/index.js

```diff
@@ -184,19 +184,175 @@ class TerserPlugin {
   }
 
   static hasAsset(commentFilename, assets) {
-    const assetFilenames = Object.keys(assets).map((assetFilename) =>
-      TerserPlugin.removeQueryString(assetFilename)
-    );
-
-    return assetFilenames.includes(
-      TerserPlugin.removeQueryString(commentFilename)
-    );
+    return assets.has(TerserPlugin.removeQueryString(commentFilename));
   }
 
   static isWebpack4() {
     return webpackVersion[0] === '4';
   }
 
+  *generateTasks(compiler, compilation, chunks, processedAssets) {
+    const existingAssets = new Set(
```

Comment:
We really do not need Set here, assets can very large, so we should avoid create Set, also avoiding Set should decrease CPU and memory usage